app/src/org/commcare/dalvik/utils/UriToFilePath.java

```diff
@@ -0,0 +1,139 @@
+package org.commcare.dalvik.utils;
+
+import android.annotation.SuppressLint;
+import android.content.ContentUris;
+import android.content.Context;
+import android.database.Cursor;
+import android.net.Uri;
+import android.os.Build;
+import android.os.Environment;
+import android.provider.DocumentsContract;
+import android.provider.MediaStore;
+
+/**
+ * Util for turning URIs into a normal file path strings Needed because new
+ * versions of Android have a content resolution abstraction layer that doesn't
+ * use vanilla file system paths.
+ *
+ * Taken from aFileChooser by Paul Burke (paulburke.co)
+ * https://github.com/iPaulPro/aFileChooser
+ * Under the Apache 2.0 license
+ */
+public class UriToFilePath {
+    /**
+     * Get a file path from a Uri. This will get the the path for Storage
+     * Access Framework Documents, as well as the _data field for the
+     * MediaStore and other file-based ContentProviders. Doesn't handle
+     * document Uri's on secondary storage.
+     *
+     * @param context The context.
+     * @param uri     The Uri to query.
+     * @return Filepath string extracted from the Uri argument. Returns null if
+     * filepath couldn't be succesfully extracted.
+     */
+    @SuppressLint("NewApi")
+    public static String getPathFromUri(final Context context, final Uri uri) {
+        final boolean isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT;
+
+        if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {
+            if (isExternalStorageDocument(uri)) {
+                final String docId = DocumentsContract.getDocumentId(uri);
+                final String[] split = docId.split(":");
+                final String type = split[0];
+
+                if ("primary".equalsIgnoreCase(type)) {
+                    return Environment.getExternalStorageDirectory() + "/" + split[1];
+                }
+                // XXX PLM: Doesn't handle non-primary volumes
```

Comment:
I'd like to add support for secondary storage volumes but can't figure out how to test it, and there is very limited info online